mentat/terminal/patched_autocomplete.py

```diff
@@ -0,0 +1,191 @@
+                self.input_widget.cursor_position,  # pyright: ignore
+            )
+
+    async def _get_completions(self) -> List[DropdownItem]:
+        text = self.input_widget.value[
+            : self.input_widget.cursor_position  # pyright: ignore
+        ]
+        message = self.stream.send(
+            text,
+            source=StreamMessageSource.CLIENT,
+            channel="completion_request",
+            command_autocomplete=self.parent.parent.command_autocomplete,  # pyright: ignore
+        )
+
+        response = await self.stream.recv(channel=f"completion_request:{message.id}")
+        completions: List[Completion] = response.data
+        return [
+            CompletionDropdownItem(
+                content=completion["content"],
+                position=completion["position"],
+                main=(
+                    completion["display"]
+                    if completion["display"] is not None
+                    else completion["content"]
+                ),
+            )
+            for completion in completions
+        ]
+
+    # These 2 functions are very similar to their synchronous versions
+    async def _async_input_cursor_position_changed(self, cursor_position: int) -> None:
+        if (
+            self.input_widget is not None  # pyright: ignore
+            and not self.history_suggester.just_moved(self.input_widget.value)
+        ):
+            matches = await self._get_completions()
+            self._mentat_sync_state(self.input_widget.value, cursor_position, matches)
+
+    async def _async_input_value_changed(self, value: str) -> None:
+        if (
+            self.input_widget is not None  # pyright: ignore
+            and not self.history_suggester.just_moved(value)
+        ):
+            matches = await self._get_completions()
+            self._mentat_sync_state(
+                value, self.input_widget.cursor_position, matches  # pyright: ignore
+            )
+
+    # Since we do the autocomplete matching on the backend, we completely change this function
+    # to not do any matching of its own
+    def _mentat_sync_state(
+        self, value: str, input_cursor_position: int, matches: List[DropdownItem]
+    ):
+        self.child.matches = matches
+        if matches:
+            self.styles.width = max(len(match.main) for match in matches) + 2
+
+        self.display = len(matches) > 0 and value != "" and self.input_widget.has_focus
+        self.cursor_home()
+        self.reposition(input_cursor_position)
+        self.child.refresh()
+
+    # By default the dropdown can only be below the input, but since our input is always at the bottom of the screen
+    # we want the opposite, so we patch this function to drop up instead of down
+    @override
+    def reposition(
+        self,
+        input_cursor_position: int | None = None,
+        scroll_target_adjust_y: int = 0,
+    ) -> None:
+        if self.input_widget is None:  # pyright: ignore
```

Comment:
Assuming we're ignoring IncompatibleOverride errors here?

Reply:
I think the one being ignored here is that self.input_widget has a type of Input, not Input | None (of course it could still be None if it isn't set, which is what we're checking for); there are a ton of pyright ignores in this file because I was mostly just copying the code from the classes I patched the code from and I didn't want to change it.